crates/subspace-block-relay/src/protocol/full_block.rs

```diff
@@ -0,0 +1,512 @@
+    ) {
+        if !self.pending_downloads.remove(&request.0) {
+            warn!(
+                target: LOG_TARGET,
+                "FullBlockRelay::on_block_download_response(): unknown request: \
+                peer = {:?}, request = {}, len = {}",
+                sender,
+                request,
+                bytes.len()
+            );
+            return;
+        }
+
+        let response_len = bytes.len();
+        let block_response = match BlockResponse::<Block>::decode(&mut bytes.as_slice()) {
+            Ok(block_response) => block_response,
+            Err(err) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "FullBlockRelay::on_block_download_response(): failed to decode response: \
+                    peer = {:?}, request = {}, len = {}, err = {:?}",
+                    sender,
+                    request,
+                    response_len,
+                    err
+                );
+                return;
+            }
+        };
+
+        // Import the downloaded block.
+        let reponse_str = format!("{block_response}");
+        self.import_block(sender, block_response.0).await;
+        trace!(
+            target: LOG_TARGET,
+            "FullBlockRelay::on_block_download_response(): {}, {}. Block downloaded/imported, \
+            response len = {}, elapsed = {:?}",
+            request,
+            reponse_str,
+            response_len,
+            elapsed,
+        );
+    }
+
+    /// Imports the block.
+    async fn import_block(&mut self, sender: PeerId, block: SignedBlock<Block>) {
+        let (header, extrinsics) = block.block.deconstruct();
+        let hash = header.hash();
+        self.import_queue.import_blocks(
+            BlockOrigin::ConsensusBroadcast,
+            vec![IncomingBlock::<Block> {
+                hash,
+                header: Some(header),
+                body: Some(extrinsics),
+                indexed_body: None,
+                justifications: block.justifications,
+                origin: Some(sender),
+                allow_missing_state: false,
+                import_existing: false,
+                state: None,
+                skip_execution: false,
+            }],
+        );
+    }
+
+    /// Sends the response to the block download request.
+    async fn send_download_response(
+        &mut self,
+        incoming: IncomingRequest,
+        request: BlockRequest<Block>,
+        response: BlockResponse<Block>,
+    ) {
+        let encoded = response.encode();
+        let encoded_len = encoded.len();
+        let outgoing = OutgoingResponse {
+            result: Ok(encoded),
+            reputation_changes: vec![],
+            sent_feedback: None,
+        };
+
+        if let Err(err) = incoming.pending_response.send(outgoing) {
+            warn!(
+                target: LOG_TARGET,
+                "FullBlockRelay::send_download_response(): {}, {}, response len = {}, err = {:?}",
```

Comment:
Just a thought, not crucial, with some more human-readable context, I feel like it could be more concise but equivalently helpful without having to manually add the function name as a prefix.

Reply:
Cleaned logs